### ISSUE TITLE
no-restricted-declarations rule

### ISSUE THREAD

**Issue by rightaway** [eslint/eslint#8395]
As discussed in https://github.com/eslint/eslint/issues/8031, a `no-restricted-declarations` rule would be useful, for example in preventing to assign to a variable named `undefined`, which could cause subtle bugs.

**Comment by vitorbal** [comment]
Hi @rightaway! It looks like the consensus in #8031 was that a `no-restricted-declarations` rule would be too specific for core and that it should be implemented as a custom rule instead.

Has anything changed since #8031 that would justify including this rule in core now?

**Comment by rightaway** [comment]
I've come across some other cases where it would be good to blacklist certain variable/argument names that lead to confusion in developer teams, or that are misused (e.g. `object`), so I think the rule would have more uses than just for preventing variable named `undefined`.

**Comment by vitorbal** [comment]
@rightaway thanks for clarifying 👍 

I think it would be helpful as well if we could come up with a good schema for this new rule. What do you currently have in mind?

Would you be willing to submit a PR for this new rule in case the proposal got approved? This is not necessary but would increase the chances of this being approved, since the team has limited availability!

**Comment by platinumazure** [comment]
I used the term `no-restricted-declarations`, but should this be about assignments or declarations? I might have misunderstood @rightaway's use case.

**Comment by not-an-aardvark** [comment]
@rightaway Could you please fill out the [rule proposal template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md)? I'm not sure I understand what this rule would do.

**Comment by rightaway** [comment]
**Please describe what the rule should do:**

It should warn when declaring anything (e.g. variables, function parameters) specified in the rule. Examples would be `undefined`, which could be problematic if `undefined` is reassigned. Or something like `object`, which is a matter of code style and not a descriptive variable name.

**What category of rule is this? (place an "X" next to just one item)**

[x] Enforces code style
[ ] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

The use of `object` here should cause warnings if `object` is specified as a restricted word.

```js
let object = myVariable
const { object } = myVariable
function myFunction(object) {}
```

**Why should this rule be included in ESLint (instead of a plugin)?**

Variable naming is a key part of writing code (both as a matter of style and preventing errors) so I think it would be a good thing to have it in ESLint.

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get enough support from the team and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after a long time tend to never do it, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.
